packages/tasit-action/src/TasitAction.test.js

```diff
@@ -84,109 +77,292 @@ describe("Contract", () => {
+      simpleStorage.setWallet(wallet);
+      simpleStorage.removeWallet();
+      expect(() => {
+        simpleStorage.setValue("hello world");
+      }).to.throw();
+    });
   });
 
-  it("should throw when subscribing without callback", async () => {
-    simpleStorage = simpleStorage.setWallet(wallet);
-    const subscription = simpleStorage.setValue("hello world");
-
-    try {
-      await subscription.on("confirmation");
-      assert(false, "subscribing without a callback");
-    } catch (e) {
-      await subscription.removeAllListeners();
-      assert(true);
-    }
-  });
+  describe("messages/transaction subscriptions tests", async () => {
+    let subscription;
 
-  it("should call a write contract method (send tx)", async () => {
-    simpleStorage = simpleStorage.setWallet(wallet);
+    beforeEach("assign a wallet to the contract", () => {
+      expect(() => {
+        simpleStorage.setWallet(wallet);
+      }).to.not.throw();
+    });
 
-    var rand = Math.floor(Math.random() * Math.floor(1000)).toString();
-    const subscription = simpleStorage.setValue(rand);
+    afterEach("waiting for message/tx confirmation", async () => {
+      if (subscription) {
+        await subscription.waitForMessage();
+        subscription.removeAllListeners();
+      }
+    });
 
-    const onMessage = async message => {
-      // message.data = Contents of the message.
-      const { data } = message;
-      const { confirmations } = data;
+    it("should throw when subscribing with invalid trigger", async () => {
+      subscription = simpleStorage.setValue("hello world");
+      return expect(subscription.on("invalid", () => {})).to.be.rejected;
+    });
 
-      if (confirmations === 7) {
-        subscription.removeAllListeners();
+    it("should throw when subscribing without callback", async () => {
+      subscription = simpleStorage.setValue("hello world");
+      return expect(subscription.on("confirmation")).to.be.rejected;
+    });
+
+    it("should call a write contract method (send tx) - imediate subscription", async () => {
+      var rand = Math.floor(Math.random() * Math.floor(1000)).toString();
+      subscription = simpleStorage.setValue(rand);
+      const fakeFn = sinon.fake();
+
+      const onMessage = async message => {
+        const { data } = message;
+        const { confirmations } = data;
+
+        if (confirmations >= 7) {
+          fakeFn();
+
+          subscription.removeAllListeners();
+
+          const value = await simpleStorage.getValue();
+          expect(value).to.equal(rand);
+
+          // FIXME: UnhandledPromiseRejectionWarning
+          //expect(1).to.equal(2);
+        }
+      };
+
+      await subscription.on("confirmation", onMessage);
+
+      await mineBlocks(simpleStorage.getProvider(), 15);
+
+      expect(fakeFn.called).to.be.true;
+    });
+
+    it("should call a write contract method (send tx) - late subscription", async () => {
```

Comment:
The name of this test isn't descriptive enough yet. What should happen when we call a write contract method and subscribe late?

Reply:
From the code in the test it looks like the listener should still fire even if it's added late. I think I wouldn't have expected this to be the behavior we'd want. Do you think it's better the way you have it here? If so, why? This approach may very well make more sense, so I'm curious to hear your thoughts.

Reply:
I've done that way because I've thought that was that you has asked to do. 
I have no strong option about that, I'll point out some pros/cons to keep that way:
Pros:
- Could be easier for devs because `confirmation` event almost never will be "stuck";
- If there is a case where the time between action and the listener is too long (I couldn't think a case such that).

Cons:
- Could be error-prone behavior, UI could alert that older action has just occurred.
- Differ of all blockchain event-listener that I know (maybe it isn't a big deal).

After all, I think that we should implement our tx listener without "late confirmation".

Reply:
Thanks for sharing those pros and cons. I agree with the conclusion you reached. Should we create an issue for that?

Reply:
https://github.com/tasitlabs/TasitSDK/issues/110

Reply:
Thanks! Resolving this.